### PR TITLE
Suppressed errors in pyright

### LINKED ISSUE #4642: [wishlist] Resolve pyright errors
## Current use of static type checkers

Cloud-init uses type annotations with Python static analyzers to check for typing errors. This increases code quality, and thereby the confidence that new releases of cloud-init will not experience unexpected runtime Python typing exceptions. Currently cloud-init includes a [mypy ci job](https://github.com/canonical/cloud-init/blob/cd35cadc5cbd952ccd17932688ff9dfef2a160c7/.github/workflows/check_format.yml#L20) to validate new contributions.

## Pyright: a more verbose alternative

[Pyright](https://github.com/microsoft/pyright) provides an alternative static type checker implementation. This checker (or sometimes Microsoft's proprietary version Pylance) is often used by developers as an LSP server (with VSCode, neovim, etc) to assist with development. However, it is much more verbose; Pyright reports many potential issues which mypy does not report. 

## Request: resolve (fix or silence) currently reported errors

This request is to resolve currently reported errors and optionally generate a configuration for Pyright which silences errors that are not helpful. Some of the reported errors may represent potential errors, and many are expected to be code paths that simply do not occur, or false positives.

Resolving this bug is expected to increase the quality of cloud-init's code, increase value of Pyright / Pylance to developers by increasing the signal to noise ratio, and additionally make it possible to run Pyright as a continuous integration test[1].

Slightly over half of Pyright errors occur in test code. Fixing test code is a lower priority than fixing runtime code (code that lives in `cloudinit/`). Many errors may be solved without deep understanding of cloud-init's code base. This is a good first issue that would be an excellent way for a new developer to become familiar with the code base.

## Steps to reproduce the problem
```
$ git clone cloud-init
$ cd cloud-init
$ python3 -m venv venv
$ . venv/bin/activate
$ pip install pyright
$ pyright cloudinit/    # runtime code (fix these first)
$ pyright tests/        # test code
```

[1] continuous integration testing is not an explicit goal of this issue - whether that is desirable is yet to be determined

### PR DESCRIPTION
**Pull request by aadityasinha-dotcom**
## Proposed Commit Message
Resolves #4642

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -59,3 +59,15 @@ ignore = [
 ]
 [tool.ruff.per-file-ignores]
 "cloudinit/cmd/main.py" = ["E402"]
+
+[tool.pyright]
```
Comment (aadityasinha-dotcom) [inline pyproject.toml:63]: Need some review on that :relaxed:

**Comment by TheRealFalcon** [comment]
@aadityasinha-dotcom , thanks for the contribution. Can you explain your changes on this PR? I don't understand why the configuration you're submitted would be a considered an improvement over the defaults.

**Comment by aadityasinha-dotcom** [comment]
Hi sorry I should mention the changes that were reflected in the description.
Before the configuration `pyright cloudinit/` was giving `556 errors` now it is giving `368 errors`. I tried to supress the warnings in the pyright.

**Comment by TheRealFalcon** [comment]
@aadityasinha-dotcom , Thanks, but the intention of #4642 was to fix the actual errors over time, rather than just mask them. What is the reasoning for suppressing the issues that this PR is suppressing? Do you believe them to be invalid in some way?

**Comment by aadityasinha-dotcom** [comment]
@TheRealFalcon yes I think you are correct. We should fix the actual errors. There was also a point "`optionally generate a configuration for Pyright which silences errors that are not helpful`". I will look into the errors again and will try to find the way to reslove them.

**Comment by TheRealFalcon** [comment]
@aadityasinha-dotcom , thanks. We aren't expecting all errors to be resolved in one go. That would be an enormous effort. The spirit of the issue was more to highlight that those errors exist, and we're always happy to reduce the number where we can. We're also ok silencing errors if the errors don't make sense or are too strict. We'd just need some justification as to why they don't make sense.

Should we close this PR as-is, or are there still changes here you want to make?

**Comment by aadityasinha-dotcom** [comment]
@TheRealFalcon yeah sure I will close this PR and will try to come up with the solution :+1:

**Pull request closed without merging**